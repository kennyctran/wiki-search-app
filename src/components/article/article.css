.article-container {
  width: 100%;
  height: fit-content;
  box-shadow: 0px 0px 3px 0px grey;
  border-radius: 3px;
  display: flex;
  /* gap: 0.5rem; */
  padding: 7px;
}

.article-title {
  word-wrap: break-word;
  width: 60%;
  /* max-width: 45%; */
  min-width: 100px;
  font-size: 1rem;
  font-weight: 300;
}

.article-title > span {
  font-size: 0.9rem;
  font-weight: 700;
}

.article-views-container {
  width: 20%;
  padding: 10px;
  display: flex;
  flex-direction: column;
  flex-wrap: nowrap;
  justify-content: center;
  /* max-width: 25%; */
  /* min-width: 65px; */
}

.article-views-container > span:first-child {
  font-size: 0.8rem;
}

.article-views-container > span:last-child {
  font-size: 0.7rem;
}

.button-container {
  display: flex;
  width: 20%;
}

button {
  width: 100%;
  font-size: 0.7rem;
  font-weight: 700;
  display: flex;
  justify-content: center;
  cursor: pointer;
  border-radius: 5px;
  border: 0.5px solid gray;
  background-color:rgb(238, 240, 234);
  transition: background-color 0.1s;
}

button:hover {
  background-color: rgb(209, 211, 212);
  transition: background-color 0.2s;
}

button:active {
  background-color: rgb(175, 175, 175);
}

button > span {
  align-self: center;
}
