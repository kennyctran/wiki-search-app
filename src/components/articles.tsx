import React from 'react';

interface IArticlesProps {
  articles: any[];
  limit: number;
}

export const Articles = (props: IArticlesProps): JSX.Element => {
  const { articles, limit } = props;
  // Possible TODO: Pagination?
  return (
    <div>
      { articles.map((article, rank) => {
        if (rank < limit) {
          return (
            <div key={rank}>
              <h2>{article.article}</h2>
              <p>views: {article.views}</p>
            </div>
          )
        }
        return null;
      })}
    </div>
  );
}
