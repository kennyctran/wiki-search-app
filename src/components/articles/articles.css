.articles-container {
  width: 50%;
  min-width: 225px;
  display: flex;
  flex-direction: column;
  gap: 16px;
}