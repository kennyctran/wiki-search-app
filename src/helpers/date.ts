import { format } from 'date-fns';

export const getYesterday = (structure: string): string => {
  const date = new Date();
  date.setDate(date.getDate() - 1);
  return format(date, structure);
}