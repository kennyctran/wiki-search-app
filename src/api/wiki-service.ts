import { get } from "./http-client";
import { format } from 'date-fns';
import { IArticle } from "../components/article/article";

export const getArticles = async (date: string): Promise<IArticle[]> => {
  try {
    const formattedDate = format(new Date(date), 'yyyy/MM/dd');
    const response = await get(`https://wikimedia.org/api/rest_v1/metrics/pageviews/top/en.wikipedia/all-access/${formattedDate}`);
    const [ items ] = response.items;
    return items.articles.slice(0, 200);
  } catch (err) {
    console.error('Could not fetch wiki articles: ', err);
    return [];
  }
}