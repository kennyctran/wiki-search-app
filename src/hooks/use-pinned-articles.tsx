import { useCallback, useState } from "react";
import { IArticle } from "../components/article/article";
import { LocalStorageKeys } from "../features/wiki-search";

interface IUsePinnedArticlesReturn {
  pinnedArticles: IArticle[];
  pinArticle: (article: IArticle) => void;
  unpinArticle: (article: IArticle) => void;
}

export const usePinnedArticles = (): IUsePinnedArticlesReturn => {
  const [articles, setArticles] = useState<IArticle[]>(() => {
    const storedArticles = localStorage.getItem(LocalStorageKeys.PinnedArticles);
    return storedArticles ? JSON.parse(storedArticles) : [];
  });

  const pinArticle = useCallback((article: IArticle) => {
    const savedArticles = [ ...articles, article ];
    setArticles(savedArticles);
    localStorage.setItem(LocalStorageKeys.PinnedArticles, JSON.stringify(savedArticles));
  }, [articles]);

  const unpinArticle = useCallback((article: IArticle) => {
    const savedArticles = articles.filter((item: IArticle) => item.article !== article.article);
    setArticles(savedArticles);
    localStorage.setItem(LocalStorageKeys.PinnedArticles, JSON.stringify(savedArticles));
  }, [articles]);

  return { pinnedArticles: articles, pinArticle, unpinArticle };
}