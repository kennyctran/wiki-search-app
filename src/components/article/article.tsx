import { HandleRemoveType, HandleSaveType } from "../articles/articles";
import './article.css';

export interface IArticle {
  article: string;
  rank: number;
  views: number;
}

export interface IArticleProps {
  article: IArticle;
  handleSave?: HandleSaveType;
  handleRemove?: HandleRemoveType;
}

export const Article = (props: IArticleProps): JSX.Element => {
  const { article, handleSave, handleRemove } = props;
  const titleWithSpaces = article.article.replace(/\_/g, ' ');
  return (
    <div key={article.rank} className="article-container">
      <h3 className="article-title"><span>{article.rank}.{' '}</span>{titleWithSpaces}</h3>
      <div className="article-views-container"><span>views:</span> <span>{article.views}</span></div>
      <div className="button-container">
        { handleSave && <button data-testid="save-button" onClick={() => handleSave(article)}><span>save</span></button> }
        { handleRemove && <button data-testid="remove-button" onClick={() => handleRemove(article)}><span>remove</span></button>}
      </div>
    </div>

  )
}