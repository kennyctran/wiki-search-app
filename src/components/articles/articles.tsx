import React from 'react';
import { Article, IArticle } from '../article/article';
import  './articles.css';

interface IArticlesProps {
  articles: any[];
  limit?: number;
  handleSave?: HandleSaveType;
  handleRemove?: HandleRemoveType;
}

export type HandleSaveType = (article: IArticle) => void;
export type HandleRemoveType = (article: IArticle) => void;

export const Articles = React.memo((props: IArticlesProps): JSX.Element => {
  const { articles, limit, handleSave, handleRemove } = props;
  return (
    <article className='articles-container'>
      { articles.map((article: IArticle, rank: number) => {
        return (!limit || rank < limit)
          ? <Article article={article} handleSave={handleSave} handleRemove={handleRemove}/>
          : null
      })}
    </article>
  );
});
