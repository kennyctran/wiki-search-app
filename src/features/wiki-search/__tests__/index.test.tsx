import { render, waitFor, screen, waitForElementToBeRemoved, fireEvent } from '@testing-library/react';
import { WikiSearch } from '../index';
import * as UsePinnedHook from '../../../hooks/use-pinned-articles';
import { IArticle } from '../../../components/article/article';
import * as http from '../../../api/wiki-service';
import { getYesterday } from '../../../helpers/date';
import { createArticleFixtures } from '../__fixtures__/articles.fixtures';

describe('Wiki Search App', () => {
  let articles: IArticle[] = createArticleFixtures(200);
  let pinArticleSpy: jest.Mock;
  let unpinArticleSpy: jest.Mock;
  let getArticlesSpy: jest.SpyInstance;

  const mockPinnedArticles = (articles: IArticle[]): void => {
    jest.spyOn(UsePinnedHook, 'usePinnedArticles').mockReturnValue({
      pinnedArticles: articles,
      pinArticle: pinArticleSpy,
      unpinArticle: unpinArticleSpy
    });
  }

  beforeEach(() => {
    pinArticleSpy = jest.fn();
    unpinArticleSpy = jest.fn();

    mockPinnedArticles([]);

    getArticlesSpy = jest.spyOn(http, 'getArticles').mockResolvedValue(articles);
  });

  it('should match snapshot with no articles', () => {
    const { container } = render(<WikiSearch />);
    jest.spyOn(http, 'getArticles').mockResolvedValue([]);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with articles', () => {
    mockPinnedArticles(createArticleFixtures(3));
    const { container } = render(<WikiSearch />);
    expect(container).toMatchSnapshot();
  });

  it('should always have a page title', async () => {
    const { queryByText, getByText } =  render(<WikiSearch />);

    expect(getByText('Most Popular Pages Search')).toBeInTheDocument();
    await waitFor(() => {
      expect(getArticlesSpy).toHaveBeenCalled();
    });
    await waitForElementToBeRemoved(() => queryByText('Loading...'));
  });

  it('should fetch from Wiki API on init and display the default number of articles', async () => {
    const { queryByText, container } = render(<WikiSearch />);

    expect(screen.getByText('Loading...')).toBeInTheDocument();

    await waitFor(() => {
      expect(getArticlesSpy).toHaveBeenCalledWith(getYesterday('yyyy/MM/dd'));
      expect(getArticlesSpy).toHaveBeenCalledTimes(1);
    });
    await waitForElementToBeRemoved(() => queryByText('Loading...'));
    expect(container.getElementsByClassName('article-container')).toHaveLength(100);
  });


  it('should show a placeholder if there are no pages after fetching', async () => {
    getArticlesSpy = jest.spyOn(http, 'getArticles').mockResolvedValue([]);

    const { queryByText } = render(<WikiSearch />);

    await waitForElementToBeRemoved(() => queryByText('Loading...'));

    expect(queryByText('Unable to find page views, please try a different date...')).toBeInTheDocument();
  });

  it('should add a pinned page if user saves it', async () => {
    const { queryByText, queryAllByTestId } = render(<WikiSearch />);

    expect(queryByText('Pinned Pages:')).toBeInTheDocument();
    expect(queryByText('no pinned pages...yet')).toBeInTheDocument();

    await waitForElementToBeRemoved(() => queryByText('Loading...'));

    const saveButton = queryAllByTestId('save-button')[0];
    saveButton.click();

    expect(pinArticleSpy).toHaveBeenCalledTimes(1);
  });


  it('should show pinned pages if saved', async () => {
    mockPinnedArticles(createArticleFixtures(2));

    const { queryByText, queryAllByTestId } = render(<WikiSearch />);
    await waitFor(() => expect(queryAllByTestId('remove-button')).toHaveLength(2));
    await waitForElementToBeRemoved(() => queryByText('Loading...'));
  });

  it('should display as many articles as possible if not reaching result limit', async () => {
    jest.spyOn(http, 'getArticles').mockResolvedValue(createArticleFixtures(9));

    const { queryByText, queryAllByTestId } = render(<WikiSearch />);
    await waitForElementToBeRemoved(() => queryByText('Loading...'));
    await waitFor(() => expect(queryAllByTestId('save-button')).toHaveLength(9));
  });

  it('should not display back to top button if no articles', async () => {
    jest.spyOn(http, 'getArticles').mockResolvedValue([]);
    const { queryByText } = render(<WikiSearch />);
    await waitForElementToBeRemoved(() => queryByText('Loading...'));
    expect(queryByText('back to top')).not.toBeInTheDocument();
  });

  it('should display back to top button if there are articles', async () => {
    const { queryByText } = render(<WikiSearch />);
    await waitForElementToBeRemoved(() => queryByText('Loading...'));
    expect(queryByText('back to top')).toBeInTheDocument();
  });


  it('should display date selector and result limiter regardless of articles', async () => {
    const { container, queryByText } = render(<WikiSearch />);

    expect(container.getElementsByClassName('date-picker-container')).toHaveLength(1);
    expect(container.getElementsByClassName('results-limiter-container')).toHaveLength(1);

    await waitForElementToBeRemoved(() => queryByText('Loading...'));

    expect(container.getElementsByClassName('date-picker-container')).toHaveLength(1);
    expect(container.getElementsByClassName('results-limiter-container')).toHaveLength(1);
  });

  it('should display the amount of articles selected by the user', async () => {
    const { container, queryByText, getByTestId } = render(<WikiSearch />);

    const changeLimitAndCheck = async (numOfArticles: number): Promise<void> => {
      fireEvent.change(getByTestId('result-limit-selector'), { target: { value: numOfArticles }});

      await waitFor(() => {
        expect(container.getElementsByClassName('article-container')).toHaveLength(numOfArticles);
      });
      expect(queryByText('no pinned pages...yet')).toBeInTheDocument();
    }

    await waitForElementToBeRemoved(() => queryByText('Loading...'));

    expect(container.getElementsByClassName('article-container')).toHaveLength(100);
    expect(queryByText('no pinned pages...yet')).toBeInTheDocument();

    await changeLimitAndCheck(25);
    await changeLimitAndCheck(200);
    await changeLimitAndCheck(50);
    await changeLimitAndCheck(75);
  });

  it('should attempt to update articles if date is changed', async () => {
    const { queryByText, getByTestId } = render(<WikiSearch />);
    await waitForElementToBeRemoved(() => queryByText('Loading...'));

    expect(getArticlesSpy).toHaveBeenCalledTimes(1);

    fireEvent.change(getByTestId('date-selector'), { target: { value: '2015-01-21' }});

    await waitFor(() => {
      expect(getArticlesSpy).toHaveBeenCalledTimes(2);
    });
  });
});