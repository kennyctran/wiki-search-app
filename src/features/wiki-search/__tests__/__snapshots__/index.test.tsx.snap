// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wiki Search App should match snapshot with articles 1`] = `
<div>
  <div
    id="wiki-search-feature"
  >
    <h1
      class="app-title"
    >
      Most Popular Pages Search
    </h1>
    <aside
      class="pinned-articles-container"
    >
      <h2>
        Pinned Pages: 
      </h2>
      <article
        class="articles-container"
      >
        <div
          class="article-container"
        >
          <h3
            class="article-title"
          >
            <span>
              1
              .
               
            </span>
            test title 0
          </h3>
          <div
            class="article-views-container"
          >
            <span>
              views:
            </span>
             
            <span>
              1
            </span>
          </div>
          <div
            class="button-container"
          >
            <button
              data-testid="remove-button"
            >
              <span>
                remove
              </span>
            </button>
          </div>
        </div>
        <div
          class="article-container"
        >
          <h3
            class="article-title"
          >
            <span>
              2
              .
               
            </span>
            test title 1
          </h3>
          <div
            class="article-views-container"
          >
            <span>
              views:
            </span>
             
            <span>
              1
            </span>
          </div>
          <div
            class="button-container"
          >
            <button
              data-testid="remove-button"
            >
              <span>
                remove
              </span>
            </button>
          </div>
        </div>
        <div
          class="article-container"
        >
          <h3
            class="article-title"
          >
            <span>
              3
              .
               
            </span>
            test title 2
          </h3>
          <div
            class="article-views-container"
          >
            <span>
              views:
            </span>
             
            <span>
              1
            </span>
          </div>
          <div
            class="button-container"
          >
            <button
              data-testid="remove-button"
            >
              <span>
                remove
              </span>
            </button>
          </div>
        </div>
      </article>
    </aside>
    <br />
    <section
      class="search-selectors-container"
    >
      <div
        class="date-picker-container"
      >
        <label
          for="date-picker"
        >
          Select date: 
        </label>
        <input
          data-testid="date-selector"
          id="date-picker"
          name="date-picker"
          type="date"
          value="2022-05-12"
        />
      </div>
      <div
        class="results-limiter-container"
      >
        <label
          for="result-limiter"
        >
          Number of Results: 
        </label>
        <select
          data-testid="result-limit-selector"
          id="result-limiter"
          name="results"
        >
          <option
            value="25"
          >
            25
          </option>
          <option
            value="50"
          >
            50
          </option>
          <option
            value="75"
          >
            75
          </option>
          <option
            selected=""
            value="100"
          >
            100
          </option>
          <option
            value="200"
          >
            200
          </option>
        </select>
      </div>
    </section>
    <section
      class="searched-articles-container"
    >
      <h3
        class="placeholder"
      >
        Loading...
      </h3>
    </section>
  </div>
</div>
`;

exports[`Wiki Search App should match snapshot with no articles 1`] = `
<div>
  <div
    id="wiki-search-feature"
  >
    <h1
      class="app-title"
    >
      Most Popular Pages Search
    </h1>
    <aside
      class="pinned-articles-container"
    >
      <h2>
        Pinned Pages: 
      </h2>
      <h3
        class="placeholder"
      >
        no pinned pages...yet
      </h3>
    </aside>
    <br />
    <section
      class="search-selectors-container"
    >
      <div
        class="date-picker-container"
      >
        <label
          for="date-picker"
        >
          Select date: 
        </label>
        <input
          data-testid="date-selector"
          id="date-picker"
          name="date-picker"
          type="date"
          value="2022-05-12"
        />
      </div>
      <div
        class="results-limiter-container"
      >
        <label
          for="result-limiter"
        >
          Number of Results: 
        </label>
        <select
          data-testid="result-limit-selector"
          id="result-limiter"
          name="results"
        >
          <option
            value="25"
          >
            25
          </option>
          <option
            value="50"
          >
            50
          </option>
          <option
            value="75"
          >
            75
          </option>
          <option
            selected=""
            value="100"
          >
            100
          </option>
          <option
            value="200"
          >
            200
          </option>
        </select>
      </div>
    </section>
    <section
      class="searched-articles-container"
    >
      <h3
        class="placeholder"
      >
        Loading...
      </h3>
    </section>
  </div>
</div>
`;
